pkg/server/server.go

```diff
@@ -517,3 +540,63 @@ func goContext(parent genericapiserver.PostStartHookContext) context.Context {
 func (s *Server) WaitForPhase1Finished() {
 	<-s.rootPhase1FinishedCh
 }
+
+func (s *Server) leaderElectAndRun(ctx context.Context) {
+	logger := klog.FromContext(ctx)
+
+	id, err := os.Hostname()
+	if err != nil {
+		logger.Error(err, "failed to determine hostname")
+		return
+	}
+
+	id = id + "_" + string(uuid.NewUUID())
+
+	rl, err := resourcelock.NewFromKubeconfig("leases", metav1.NamespaceSystem, "kcp-controllers",
+		resourcelock.ResourceLockConfig{
+			Identity: id,
+		},
+		s.ExtraConfig.LeaderElectionConfig,
+		time.Second*30,
+	)
+
+	if err != nil {
+		logger.Error(err, "failed to take resourcelock")
+		return
+	}
+
+	logger = logger.WithValues("identity", id)
+	logger.Info("starting leader election")
+
+	leaderelection.RunOrDie(ctx, leaderelection.LeaderElectionConfig{
+		Lock:          rl,
+		LeaseDuration: time.Second * 60,
+		RenewDeadline: time.Second * 5,
+		RetryPeriod:   time.Second * 2,
+		Callbacks: leaderelection.LeaderCallbacks{
+			OnStartedLeading: func(ctx context.Context) {
+				logger.Info("started leading")
+
+				for controllerName, controller := range s.controllers {
+					logger := logger.WithValues("controller", controllerName)
+					go s.runController(ctx, controller, logger)
+				}
+			},
+			OnStoppedLeading: func() {
+				logger.Info("stopped leading")
+			},
+			OnNewLeader: func(current_id string) {
+				if current_id == id {
+					// we already log when we start leading, no reason to also log something here.
+					return
+				}
+
+				logger.WithValues("new-leader", current_id).Info("new leader is someone else")
+			},
+		},
+		WatchDog: leaderelection.NewLeaderHealthzAdaptor(time.Second * 5),
+		Name:     "kcp-controllers",
+	})
+
+	logger.Info("leader election has stopped")
```

Comment:
`RunOrDie` exits either when some error occurs (through a `panic`), the `ctx` is cancelled or if the lease couldn't be renewed. If lease couldn't be renewed, we exit this function entirely and don't re-enter it again.

Reply:
This is a good point - any thoughts on what should happen in that case? We could let the kcp process crash so it gets restarted or put this in a loop, but I'm anxious to just let it keep crashing again and again if there is a bigger problem.

Reply:
I have chosen to implement a continuous loop around the leader election, so that if a kcp instance loses the leader role it will restart `RunOrDie` (until the overarching context is canceled).